File: C++/increasingBST.cpp
/**
 * @File    : increasingBST.cpp
 * @Brief   : 递增顺序搜索树
 * @Link    : https://leetcode-cn.com/problems/increasing-order-search-tree/
 * @Author  : Wei Li
 * @Date    : 2021-08-02
*/

#include <iostream>
#include <vector>

// Definition for a binary tree node.
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    void inorder(TreeNode *node, std::vector<int> &res)
    {
        if (node == nullptr)
        {
            return;
        }
        inorder(node->left, res);
        res.push_back(node->val);
        inorder(node->right, res);
    }

    TreeNode *increasingBST(TreeNode *root)
    {
        std::vector<int> res;
        inorder(root, res);

        TreeNode *dummyNode = new TreeNode(-1);
        TreeNode *currNode = dummyNode;
        for (int value : res)
        {
            currNode->right = new TreeNode(value);
            currNode = currNode->right;
        }
        return dummyNode->right;
    }
};

// ----------------------------
int main(int argc, char **argv)
{

    return 0;
}

File: Python/increasingBST.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: increasingBST.py
@Function: 递增顺序搜索树
@Link : https://leetcode-cn.com/problems/increasing-order-search-tree/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-08-02
"""

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution(object):
    def increasingBST(self, root):
        dummy = TreeNode(-1)
        self.prev = dummy
        self.inOrder(root)
        return dummy.right

    def inOrder(self, root):
        if not root:
            return None
        self.inOrder(root.left)
        root.left = None
        self.prev.right = root
        self.prev = root
        self.inOrder(root.right)


# ------------------------------
if __name__ == "__main__":
    #  test on LeetCode online.
    pass
